students/javarb/sessions/2019/may/Session_34-35-Serialization/Session_34-35-Serialization.md

```diff
@@ -0,0 +1,225 @@
+System.out.println("Gson = " + gsonMean + " ms");
+//...
+```
+
+Following code chunk, performs serialization returning miliseconds time benchmark in this example for Gson serialization library:
+```java
+private static long benchmarkGson(List<Person> persons) {
+	long begin = System.currentTimeMillis();
+	new Gson().toJson(persons);
+	return System.currentTimeMillis() - begin;
+}
+```
+
+Finally, method on charge of return the mean can be seen following:
+```java
+private long mean(List<Long> benchmarks) {
+	long sum = 0;  
+	for (Long benchmark : benchmarks) {
+		sum += benchmark;
+	}
+	return sum / benchmarks.size();
+}
+```
+
+A for loop was used in order to add up benchmarks in order to after get the final mean time benchmarked. 
+
+**Note:** As a side note, there is another way to iterate by using `foreach()` lambda funcion:
+```java
+benchmarks.forEach(b -> sum +=b);
+```
+As can be seen, this kind of function is practical to use but isn't possible to modify the elements we are iterating over, so we cannot use it here.
+
+
+#### Part 2 - Benchmark project in OOP
+
+After code was running, we saw code could be improved through asimilation of OOP concepts which allow us to include any library we want to our benchmark without need to modify our main funcions and code. So instead of that, we could add another serialization library just by including the library on `build.gradle`, create the corresponding class for that implementation,  and instantiating/ executing library's `benchmark()` method.
+
+So, following explanation about our OOP code.
+
+`Serialization.java` continues being main class whicn instantiate each kind of benchmark:
+
+```java
+public class Serialization {
+    public static void main(String[] args) {
+        new JacksonBenchmark().benchmark(150, 200000);
+        new GsonBenchmark().benchmark(150, 200000);
+        new FastjsonBenchmark().benchmark(150, 200000);
+    }
+}
+```
+
+Each type of bechmark inherit from abstract class `Benchmark`. In that class,  `bechmark()` contructor  method is implemented, altogether with `mean()` and `runBenchmark()`, the latter is defined as an abstract method that is implemented in each subclass:
+
+```java
+public abstract class Benchmark {
+    public void benchmark(int samplesize, int objects) {
+        List<Person> persons = new ArrayList<>();
+
+        for (int i = 0; i < objects; i++) {
+            Person person = new Person();
+            person.setId(i);
+            person.setName("Name_" + i);
+            person.setAge(i % 100);
+            persons.add(i, person);
+        }
+
+        List<Long> results = new ArrayList<>();
+
+        for (int i = 0; i < samplesize; i++) {
+            results.add(runBenchmark(persons));
+        }
+
+        long mean = mean(results);
+        System.out.println("Result of " + getClass().getSimpleName() + " = "+ mean + " ms");
+
+    }
+
+    protected abstract Long runBenchmark(List<Person> persons);
```

Comment:
```suggestion
    protected abstract Long runBenchmark(List<Person> people);
```

Reply:
In the future @perrymant, you can probably get away with a single comment suggesting that we rename `persons` -> `people`. Saves you a fair bit of work, and makes the review comments less daunting :)